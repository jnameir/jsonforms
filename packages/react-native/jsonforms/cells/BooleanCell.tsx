import {
    CellProps,
    isBooleanControl,
    RankedTester,
    rankWith,
} from "@jsonforms/core";
import { withJsonFormsCellProps } from "@jsonforms/react";
import React from "react";
import { Checkbox } from "react-native-material-ui";

export const BooleanCell = (props: CellProps) => {
    const { data, enabled, path, handleChange } = props;

    const onChange = () => {
        handleChange(path, !data);
    };

    return (
        <Checkbox
            label=""
            value={path}
            checked={!!data}
            onCheck={onChange}
            disabled={!enabled}
        />
    );
};

/**
 * Default tester for boolean controls.
 * @type {RankedTester}
 */
export const booleanCellTester: RankedTester = rankWith(2, isBooleanControl);

export default withJsonFormsCellProps(BooleanCell);
