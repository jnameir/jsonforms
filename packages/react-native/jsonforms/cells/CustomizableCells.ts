export { BooleanCell } from "./BooleanCell";
export { DateCell } from "./DateCell";
export { DateTimeCell } from "./DateTimeCell";
export { EnumCell } from "./EnumCell";
export { MultiEnumCell } from "./MultiEnumCell";
export { IntegerCell } from "./IntegerCell";
export { NumberCell } from "./NumberCell";
export { NumberFormatCell } from "./NumberFormatCell";
export { TextCell } from "./TextCell";
export { TextAreaCell } from "./TextAreaCell";
export { TimeCell } from "./TimeCell";
