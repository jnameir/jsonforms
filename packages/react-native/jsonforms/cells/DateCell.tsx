import {
    CellProps,
    isDateControl,
    RankedTester,
    rankWith,
} from "@jsonforms/core";
import { withJsonFormsCellProps } from "@jsonforms/react";
import React from "react";
import { DatePickerAndroid } from "react-native";
import { TextField } from "react-native-material-textfield";
import { Button } from "react-native-material-ui";
import { twoDigits } from "../util";

export const DateCell = (props: CellProps) => {
    const { data, path, handleChange } = props;
    const formattedDate = data
        ? `${data.split("-")[2]}.${data.split("-")[1]}.${data.split("-")[0]}`
        : "";

    const onChange = async () => {
        try {
            const date = await DatePickerAndroid.open({});

            if (date.action !== DatePickerAndroid.dismissedAction) {
                handleChange(
                    path,
                    `${date.year}-${twoDigits(date.month + 1)}-${twoDigits(
                        date.day,
                    )}`,
                );
            }
        } catch ({ code, message }) {
            console.log("wrong date format");
        }
    };

    return (
        <>
            <TextField value={formattedDate} editable={false} />
            <Button onPress={onChange} text="Pick date" />
        </>
    );
};

/**
 * Default tester for date controls.
 * @type {RankedTester}
 */
export const dateCellTester: RankedTester = rankWith(2, isDateControl);

export default withJsonFormsCellProps(DateCell);
