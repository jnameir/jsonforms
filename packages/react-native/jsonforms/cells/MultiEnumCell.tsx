import {
    and,
    EnumCellProps,
    enumToEnumOptionMapper,
    RankedTester,
    rankWith,
    schemaMatches,
    schemaTypeIs,
    uiTypeIs
} from "@jsonforms/core";
import { withJsonFormsEnumCellProps } from "@jsonforms/react";
import React from "react";
import { StyleSheet, View } from "react-native";
import SectionedMultiSelect from "react-native-sectioned-multi-select";
import Icons from "react-native-vector-icons/MaterialIcons";


export const MultiEnumCell = (props: Partial<EnumCellProps>) => {
    const { data, path, schema, handleChange, visible, enabled, errors } =
        props;

    const onChange = (value) => {
        handleChange(path, value?.length > 0 ? value : undefined);
    };

    let items = [];

    if (schema.items && "enum" in schema.items) {
        items = schema.items.enum;
    } else {
        return <></>;
    }

    const options = items.map((e) => enumToEnumOptionMapper(e));
    const preparedOptions = [{ id: 0, name: schema.title, children: [] }];

    options.forEach((e) =>
        preparedOptions[0].children.push({
            id: e.value,
            name: e.label,
        }),
    );

    const showBorder = visible
        ? errors
            ? { style: { ...styles.errorStyle } }
            : { style: { ...styles.contentBorder } }
        : null;

    /** workaround for fixing the wrong type definition of the 3rd party package react-native-sectioned-multi-select */
    const Icon = Icons as unknown as JSX.Element;

    return (
        <View style={styles.container}>
            <View {...showBorder} data-testid="MultiEnumCell-container">
                <SectionedMultiSelect
                    icons={multiSelectStyles.icons}
                    IconRenderer={Icon}
                    uniqueKey="id"
                    subKey="children"
                    items={preparedOptions}
                    selectText="Bitte auswählen..."
                    showDropDowns={false}
                    readOnlyHeadings={true}
                    hideSearch={false}
                    searchPlaceholderText="Suche"
                    confirmText="Übernehmen"
                    selectedText="ausgewählt"
                    selectedItems={data || []}
                    onSelectedItemsChange={onChange}
                    disabled={!enabled}
                    styles={multiSelectStyles.styles}
                />
            </View>
        </View>
    );
};
/**
 * Default tester for multi select enum controls.
 * @type {RankedTester}
 */

const isMultiEnumControl = and(
    uiTypeIs("Control"),
    schemaTypeIs("array"),
    schemaMatches(
        (schema) =>
            Object.prototype.hasOwnProperty.call(schema, "items") &&
            Object.prototype.hasOwnProperty.call(schema.items, "enum"),
    ),
);

export const multiEnumCellTester: RankedTester = rankWith(
    2,
    isMultiEnumControl,
);

export default withJsonFormsEnumCellProps(MultiEnumCell);

const styles = StyleSheet.create({
    container: {
        width: "50%",
        margin: 5,
    },
    contentBorder: {
        borderWidth: 1,
        borderColor: "darkgrey",
    },
    errorStyle: {
        borderWidth: 1,
        borderColor: "red",
    },
});

const multiSelectStyles = {
    icons: {
        search: {
            name: "search", // search input
            size: 24,
        },
        arrowUp: {
            name: "keyboard-arrow-up", // dropdown toggle
            size: 22,
        },
        arrowDown: {
            name: "keyboard-arrow-down", // dropdown toggle
            size: 22,
        },
        selectArrowDown: {
            name: "keyboard-arrow-down", // select
            size: 24,
        },
        close: {
            name: "close", // chip close
            size: 16,
        },
        check: {
            name: "check", // selected item
            size: 16,
        },
        cancel: {
            name: "cancel", // cancel button
            size: 18,
        },
    },
    styles: {
        selectedItem: {
            backgroundColor: "rgba(0,0,0,0.1)",
        },
        selectedSubItem: {
            backgroundColor: "rgba(0,0,0,0.1)",
        },
        button: {
            backgroundColor: "green",
        },
    },
};
