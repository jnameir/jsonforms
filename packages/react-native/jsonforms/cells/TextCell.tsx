import {
    CellProps,
    isStringControl,
    RankedTester,
    rankWith,
} from "@jsonforms/core";
import { withJsonFormsCellProps } from "@jsonforms/react";
import merge from "lodash/merge";
import React from "react";
import { TextField } from "react-native-material-textfield";

export const TextCell = (props: CellProps) => {
    const { config, data, enabled, uischema, schema, path, handleChange } =
        props;

    const onChange = (value: string) => {
        handleChange(path, value);
    };

    const { maxLength } = schema;
    const appliedUiSchemaOptions = merge({}, config, uischema.options);

    return (
        <TextField
            value={data || ""}
            onChangeText={onChange}
            secureTextEntry={
                appliedUiSchemaOptions.format === "password" ? true : false
            }
            editable={enabled}
            autoFocus={appliedUiSchemaOptions.focus}
            maxLength={appliedUiSchemaOptions.restrict ? maxLength : undefined}
            label=""
        />
    );
};

/**
 * Default tester for text-based/string controls.
 * @type {RankedTester}
 */
export const textCellTester: RankedTester = rankWith(1, isStringControl);

export default withJsonFormsCellProps(TextCell);
