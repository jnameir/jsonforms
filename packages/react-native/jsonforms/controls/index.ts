import InputControl, { inputControlTester } from "./InputControl";
import RadioGroupControl, {
    radioGroupControlTester,
} from "./RadioGroupControl";
import ConditionalControl, {
    conditionalControlTester,
} from "./ConditionalControl";

export {
    InputControl,
    inputControlTester,
    RadioGroupControl,
    radioGroupControlTester,
    ConditionalControl,
    conditionalControlTester,
};
