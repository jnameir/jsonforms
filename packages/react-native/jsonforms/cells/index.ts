import BooleanCell, { booleanCellTester } from "./BooleanCell";
import * as Customizable from "./CustomizableCells";
import DateCell, { dateCellTester } from "./DateCell";
import DateTimeCell, { dateTimeCellTester } from "./DateTimeCell";
import EnumCell, { enumCellTester } from "./EnumCell";
import MultiEnumCell, { multiEnumCellTester } from "./MultiEnumCell";
import IntegerCell, { integerCellTester } from "./IntegerCell";
import NumberCell, { numberCellTester } from "./NumberCell";
import NumberFormatCell, { numberFormatCellTester } from "./NumberFormatCell";
import TextAreaCell, { textAreaCellTester } from "./TextAreaCell";
import TextCell, { textCellTester } from "./TextCell";
import TimeCell, { timeCellTester } from "./TimeCell";

export {
    BooleanCell,
    booleanCellTester,
    DateCell,
    dateCellTester,
    DateTimeCell,
    dateTimeCellTester,
    EnumCell,
    enumCellTester,
    MultiEnumCell,
    multiEnumCellTester,
    IntegerCell,
    integerCellTester,
    NumberCell,
    numberCellTester,
    NumberFormatCell,
    numberFormatCellTester,
    TextCell,
    textCellTester,
    TextAreaCell,
    textAreaCellTester,
    TimeCell,
    timeCellTester,
};
export { Customizable };
