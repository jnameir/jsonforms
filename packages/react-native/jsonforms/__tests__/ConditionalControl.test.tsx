import { ControlElement, ControlProps, JsonSchema } from "@jsonforms/core";
import { VerticalLayout } from "@jsonforms/core/src/models/uischema";
import { DispatchCell, JsonForms } from "@jsonforms/react";
import React from "react";
import { Picker } from "react-native";
import { TextField } from "react-native-material-textfield";
import { RadioButton } from "react-native-material-ui";
import SectionedMultiSelect from "react-native-sectioned-multi-select";
import renderer, { act } from "react-test-renderer";
import { wait } from "../../api/wait";
import { MultiEnumCell } from "../cells/MultiEnumCell";
import { TextCell } from "../cells/TextCell";
import { ConditionalControl } from "../controls/ConditionalControl";
import { RNCells, RNRenderers } from "../index";

const multiSelectPath = "#/properties/nutritionalStatus";
const multiSelectLabel = "Ernährungszustand";

const multiSelectSchema = {
    title: multiSelectLabel,
    type: "array",
    items: {
        type: "string",
        enum: ["gut", "Adipositas", "Kachexie", "Exsikkose"],
    },
};

const multiSelectUISchema: ControlElement = {
    type: "Control",
    label: multiSelectLabel,
    scope: multiSelectPath,
    options: {
        format: "conditional",
        type: "enum",
        content: {
            gut: ["gut"],
            "auffällig, weil...": ["Adipositas", "Kachexie", "Exsikkose"],
        },
        nonConditionalLabel: "gut",
    },
};

const multiAndSingleSelectPath = "#/properties/treatmentByNewPhysician";
const multiAndSingleSelectLabel =
    "Erfolgt die Behandlung durch den neuen Arzt?";

const multiAndSingleSelectSchema = {
    title: multiAndSingleSelectLabel,
    type: ["array", "string"],
    items: {
        type: "string",
        enum: [
            "kein Akutbedarf",
            "kein Termin verfügbar",
            "Behandlung durch neuen Arzt nicht gewünscht",
            "Sonstiges",
            "unregelmäßig",
            "regelmäßig",
        ],
    },
};

const multiAndSingleSelectUISchema: ControlElement = {
    type: "Control",
    label: multiAndSingleSelectLabel,
    scope: multiAndSingleSelectPath,
    options: {
        format: "conditional",
        type: "enum",
        singleSelectLabels: ["ja"],
        content: {
            ja: ["unregelmäßig", "regelmäßig"],
            nein: [
                "kein Akutbedarf",
                "kein Termin verfügbar",
                "Behandlung durch neuen Arzt nicht gewünscht",
                "Sonstiges",
            ],
        },
    },
};

const textFieldPath = "#/properties/otherAbnormalities";
const textFieldLabel = "Sonstige Auffälligkeiten?";

const textFieldSchema = {
    title: textFieldLabel,
    type: "string",
};

const textFieldUISchema: ControlElement = {
    type: "Control",
    label: textFieldLabel,
    scope: textFieldPath,
    options: {
        format: "conditional",
        type: "string",
        content: {
            ja: "Bitte ausfüllen",
            nein: "",
        },
        nonConditionalLabel: "nein",
    },
};

describe("JSONForms: ConditionalControl unit test", () => {
    const getProps = (
        conditionalSchema: "multiselect" | "textfield" | "multiAndSingleSelect",
    ): ControlProps => {
        let path;
        let schema;
        let uiSchema;
        let label;

        if (conditionalSchema === "multiselect") {
            path = multiSelectPath;
            schema = multiSelectSchema;
            uiSchema = multiSelectUISchema;
            label = multiSelectLabel;
        }

        if (conditionalSchema === "textfield") {
            path = textFieldPath;
            schema = textFieldSchema;
            uiSchema = textFieldUISchema;
            label = textFieldLabel;
        }

        if (conditionalSchema === "multiAndSingleSelect") {
            path = multiAndSingleSelectPath;
            schema = multiAndSingleSelectSchema;
            uiSchema = multiAndSingleSelectUISchema;
            label = multiAndSingleSelectLabel;
        }

        return {
            data: "",
            path,
            schema,
            uischema: uiSchema,
            visible: true,
            enabled: true,
            label,
            errors: "",
            id: path,
            handleChange: () => null,
            rootSchema: {},
        };
    };

    const renderControl = (properties: ControlProps) => {
        const root = renderer.create(<ConditionalControl {...properties} />);
        const instance = root.root;

        return { root, instance };
    };

    test("success snapshot neutral", () => {
        const props = getProps("multiselect");
        const { root } = renderControl(props);
        expect(root.toJSON()).toMatchSnapshot();
    });

    test("multiselect: successfully render JSONForms DispatchedCell when conditional option is selected", () => {
        const props = getProps("multiselect");
        const { instance } = renderControl(props);
        const radioButtons = instance.findAllByType(RadioButton);
        radioButtons[1].props.onSelect("auffällig, weil...");
        const dispatchedCells = instance.findAllByType(DispatchCell);
        expect(dispatchedCells.length).toEqual(1);
    });

    test("multiselect: successfully unmount JSONForms DispatchedCell when non-conditional option is selected", () => {
        const props = getProps("multiselect");
        const { instance } = renderControl(props);
        const radioButtons = instance.findAllByType(RadioButton);
        radioButtons[0].props.onSelect("gut");
        const dispatchedCells = instance.findAllByType(DispatchCell);
        expect(dispatchedCells.length).toEqual(0);
    });

    test("multiAndSingleselect: successfully render JSONForms DispatchedCell when conditional option is selected", () => {
        const props = getProps("multiAndSingleSelect");
        const { instance } = renderControl(props);
        const radioButtons = instance.findAllByType(RadioButton);
        radioButtons[1].props.onSelect("nein");
        const dispatchedCells = instance.findAllByType(DispatchCell);
        expect(dispatchedCells.length).toEqual(1);
    });

    test("multiAndSingleSelect: successfully render JSONForms DispatchedCell when conditional option is selected", () => {
        const props = getProps("multiAndSingleSelect");
        const { instance } = renderControl(props);
        const radioButtons = instance.findAllByType(RadioButton);
        radioButtons[0].props.onSelect("ja");
        const dispatchedCells = instance.findAllByType(DispatchCell);
        expect(dispatchedCells.length).toEqual(1);
    });

    test("textfield: successfully render JSONForms DispatchedCell when conditional option is selected", () => {
        const props = getProps("textfield");
        const { instance } = renderControl(props);
        const radioButtons = instance.findAllByType(RadioButton);
        radioButtons[0].props.onSelect("ja");
        const dispatchedCells = instance.findAllByType(DispatchCell);
        expect(dispatchedCells.length).toEqual(1);
    });

    test("textfield: successfully unmount JSONForms DispatchedCell when non-conditional option is selected", () => {
        const props = getProps("textfield");
        const { instance } = renderControl(props);
        const radioButtons = instance.findAllByType(RadioButton);
        radioButtons[1].props.onSelect("nein");
        const dispatchedCells = instance.findAllByType(DispatchCell);
        expect(dispatchedCells.length).toEqual(0);
    });
});

describe("JSONForms: ConditionalControl integration test", () => {
    const schema: JsonSchema = {
        type: "object",
        properties: {
            nutritionalStatus: multiSelectSchema,
            otherAbnormalities: textFieldSchema,
            treatmentByNewPhysician: multiAndSingleSelectSchema,
        },
        required: [
            "nutritionalStatus",
            "otherAbnormalities",
            "treatmentByNewPhysician",
        ],
    };

    const uiSchema: VerticalLayout = {
        type: "VerticalLayout",
        elements: [
            multiSelectUISchema,
            textFieldUISchema,
            multiAndSingleSelectUISchema,
        ],
    };

    const fakeHandlerChange = jest.fn();

    function renderControl(initialData = {}) {
        const JSONForms = class JSONForms extends React.Component {
            render() {
                return (
                    <JsonForms
                        schema={schema}
                        uischema={uiSchema}
                        data={initialData}
                        renderers={RNRenderers}
                        cells={RNCells}
                        onChange={fakeHandlerChange}
                    />
                );
            }
        };
        const root = renderer.create(<JSONForms />);
        const instance = root.root;

        return { root, instance };
    }

    test("multiselect: successfully render on condition and select items", async () => {
        const { instance } = renderControl();
        const radioButtons = instance.findAllByType(RadioButton);

        /** select the conditional option for the multiselect component to be rendered */
        act(() => radioButtons[1].props.onSelect("auffällig, weil..."));

        const valuesToSelect = ["Kachexie", "Exsikkose"];
        const selectables = instance.findAllByType(SectionedMultiSelect);
        act(() => selectables[0].props.onSelectedItemsChange(valuesToSelect));

        expect(selectables[0].props.selectedItems).toEqual(valuesToSelect);
        await act(() => wait(100));
    });

    test("multiselect: successfully show error if all items unselected", () => {
        const { instance } = renderControl();
        const radioButtons = instance.findAllByType(RadioButton);

        /** select the conditional option for the multiselect component to be rendered */
        act(() => radioButtons[1].props.onSelect("auffällig, weil..."));

        const valuesToSelect = [];
        const selectables = instance.findAllByType(SectionedMultiSelect);
        act(() => selectables[0].props.onSelectedItemsChange(valuesToSelect));

        const multiEnumCells = instance.findAllByType(MultiEnumCell);

        expect(multiEnumCells[0].props.errors).toEqual(
            "is a required property",
        );
    });

    test("multiselect: successfully delete data when conditional component is unmounted", () => {
        const { instance } = renderControl();
        const radioButtons = instance.findAllByType(RadioButton);

        /** select the conditional option for the multiselect component to be rendered */
        act(() => radioButtons[1].props.onSelect("auffällig, weil..."));

        const valuesToSelect = ["Kachexie", "Exsikkose"];
        const selectables = instance.findAllByType(SectionedMultiSelect);
        act(() => selectables[0].props.onSelectedItemsChange(valuesToSelect));

        /** unmount component by selecting the non-conditional top-level option */
        act(() => radioButtons[0].props.onSelect("gut"));

        const conditionalControls = instance.findAllByType(ConditionalControl);

        expect(conditionalControls[0].props.data).toEqual(["gut"]);
        expect(conditionalControls[0].props.data.length).toEqual(1);
    });

    test("multiAndSingleSelect: successfully render multiselect on condition and select items", async () => {
        const { instance } = renderControl();
        const radioButtons = instance.findAllByType(RadioButton);

        /** select the conditional option for the multiselect component to be rendered */
        act(() => radioButtons[5].props.onSelect("nein"));

        const valuesToSelect = [
            "kein Akutbedarf",
            "Behandlung durch neuen Arzt nicht gewünscht",
        ];

        const selectables = instance.findAllByType(SectionedMultiSelect);
        act(() => selectables[0].props.onSelectedItemsChange(valuesToSelect));

        expect(selectables[0].props.selectedItems).toEqual(valuesToSelect);
        await act(() => wait(100));
    });

    test("multiAndSingleSelect: successfully render singleSelect on condition and select item", async () => {
        const { instance } = renderControl();
        const radioButtons = instance.findAllByType(RadioButton);

        /** select the conditional option for the singleSelect component to be rendered */
        act(() => radioButtons[4].props.onSelect("ja"));

        const valueToSelect = "regelmäßig";
        const selectables = instance.findAllByType(Picker);

        act(() => selectables[0].props.onValueChange(valueToSelect));

        expect(selectables[0].props.selectedValue).toEqual(valueToSelect);
        await act(() => wait(100));
    });

    test("multiAndSingleSelect: successfully show error if item unselected (singleSelect)", () => {
        const { instance } = renderControl();
        const radioButtons = instance.findAllByType(RadioButton);

        /** select the conditional option for the singleSelect component to be rendered */
        act(() => radioButtons[4].props.onSelect("ja"));

        const conditionalControls = instance.findAllByType(ConditionalControl);

        expect(conditionalControls[2].props.errors).toEqual(
            "is a required property",
        );
    });

    test("multiAndSingleSelect: successfully show error if item unselected (multiSelect)", () => {
        const { instance } = renderControl();
        const radioButtons = instance.findAllByType(RadioButton);

        /** select the conditional option for the singleSelect component to be rendered */
        act(() => radioButtons[4].props.onSelect("nein"));

        const conditionalControls = instance.findAllByType(ConditionalControl);

        expect(conditionalControls[2].props.errors).toEqual(
            "is a required property",
        );
    });

    test("textfield: successfully render on condition and enter text", () => {
        const { instance } = renderControl();
        const radioButtons = instance.findAllByType(RadioButton);

        /** select the conditional option for the textfield component to be rendered */
        act(() => radioButtons[2].props.onSelect("ja"));

        const textToEnter = "entered text";
        const textFields = instance.findAllByType(TextField);
        act(() => textFields[0].props.onChangeText(textToEnter));

        expect(textFields[0].props.value).toEqual(textToEnter);
    });

    test("textfield: successfully show error if no options selected", () => {
        const { instance } = renderControl();
        const conditionalControls = instance.findAllByType(ConditionalControl);
        expect(conditionalControls[1].props.errors).toEqual(
            "is a required property",
        );
    });

    test("textfield: successfully show error if rendered on condition but no text entered", async () => {
        const { instance } = renderControl();
        const radioButtons = instance.findAllByType(RadioButton);

        /** select the conditional option for the textfield component to be rendered */
        act(() => radioButtons[2].props.onSelect("ja"));
        const textCells = instance.findAllByType(TextCell);

        expect(textCells[0].props.errors).toEqual("is a required property");
    });

    test("textfield: successfully delete data when conditional component is unmounted", () => {
        const { instance } = renderControl();
        const radioButtons = instance.findAllByType(RadioButton);

        /** select the conditional option for the textfield component to be rendered */
        act(() => radioButtons[2].props.onSelect("ja"));

        const textFields = instance.findAllByType(TextField);
        act(() => textFields[0].props.onChangeText("entered text"));

        /** unmount component by selecting the non-conditional top-level option */
        act(() => radioButtons[3].props.onSelect("nein"));

        const conditionalControls = instance.findAllByType(ConditionalControl);

        expect(conditionalControls[1].props.data).toEqual("");
    });

    test("successfully preload data", () => {
        const selectedValues = ["Kachexie", "Exsikkose"];
        const enteredText = "this is a test";

        const { instance } = renderControl({
            nutritionalStatus: selectedValues,
            otherAbnormalities: enteredText,
        });

        const selectables = instance.findAllByType(SectionedMultiSelect);
        expect(selectables[0].props.selectedItems).toEqual(selectedValues);

        const textFields = instance.findAllByType(TextField);
        expect(textFields[0].props.value).toEqual(enteredText);
    });

    test("successfully preload data multiAndSingleSelect - singleselect", () => {
        const selectedValue = "regelmäßig";

        const { instance } = renderControl({
            treatmentByNewPhysician: selectedValue,
        });

        const selectables = instance.findAllByType(Picker);

        expect(selectables[0].props.selectedValue).toEqual(selectedValue);

        /** The new top-level conditional option is dynamically assigned based on to the initial preloaded values */
        const radioButtons = instance.findAllByType(RadioButton);
        expect(radioButtons[4].props.checked).toEqual(true);
    });

    test("successfully preload data multiAndSingleSelect - multiselect", () => {
        const selectedValues = ["kein Termin verfügbar", "kein Akutbedarf"];

        const { instance } = renderControl({
            treatmentByNewPhysician: selectedValues,
        });

        const selectables = instance.findAllByType(SectionedMultiSelect);

        expect(selectables[0].props.selectedItems).toEqual(selectedValues);

        /** The new top-level conditional option is dynamically assigned based on to the initial preloaded values */
        const radioButtons = instance.findAllByType(RadioButton);
        expect(radioButtons[5].props.checked).toEqual(true);
    });

    test("successfully preload data and assign the correct top-level label if there are more than one conditionally rendered multiselect questions", () => {
        /** Add additional values to the schema so that it looks like:
         {
            "type": "string",
            "enum": ["Adipositas", "Kachexie", "Exsikkose", "additionalValue1", "additionalValue2", "additionalValue3"],
         }
         * */

        schema.properties.nutritionalStatus.items["enum"].push(
            "additionalValue1",
        );
        schema.properties.nutritionalStatus.items["enum"].push(
            "additionalValue2",
        );
        schema.properties.nutritionalStatus.items["enum"].push(
            "additionalValue3",
        );

        /** Assign the new values to a new conditional option in the UI schema:
         content: {
            gut: [],
            "auffällig, weil...": ["Adipositas", "Kachexie", "Exsikkose"],
            "additionalOption": ["additionalValue1", "additionalValue2", "additionalValue3"],
         }
         * */

        uiSchema.elements[0].options.content["additionalOption"] = [
            "additionalValue1",
            "additionalValue2",
            "additionalValue3",
        ];

        const selectedValues = ["additionalValue1", "additionalValue2"];

        const { instance } = renderControl({
            nutritionalStatus: selectedValues,
        });

        const selectables = instance.findAllByType(SectionedMultiSelect);
        expect(selectables[0].props.selectedItems).toEqual(selectedValues);

        /** The new top-level conditional option is dynamically assigned based on to the initial preloaded values */
        const radioButtons = instance.findAllByType(RadioButton);
        expect(radioButtons[2].props.checked).toEqual(true);
    });
});
