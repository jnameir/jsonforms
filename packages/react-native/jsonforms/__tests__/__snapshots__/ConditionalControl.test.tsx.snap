// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONForms: ConditionalControl unit test success snapshot neutral 1`] = `
<View
  style={
    Object {
      "display": "flex",
      "padding": 5,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 16,
      }
    }
  >
    Ernährungszustand
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onLayout={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      pointerEvents="box-only"
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
          },
          undefined,
          Object {
            "flex": 0,
            "minWidth": 200,
          },
        ]
      }
      testID={null}
    >
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "rgb(0, 0, 0)",
              "borderRadius": 21.6,
              "height": 43.2,
              "left": 2.3999999999999986,
              "opacity": undefined,
              "position": "absolute",
              "top": 2.3999999999999986,
              "transform": Array [
                Object {
                  "scale": undefined,
                },
              ],
              "width": 43.2,
              "zIndex": 1,
            }
          }
        />
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "height": 48,
                "justifyContent": "center",
                "width": 48,
              },
              Object {
                "height": 48,
                "width": 48,
              },
              undefined,
            ]
          }
        >
          <Text />
        </View>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "flex": 1,
              "marginLeft": 20,
            },
            undefined,
            undefined,
            false,
          ]
        }
      >
        gut
      </Text>
      <View
        pointerEvents="none"
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "overflow": "hidden",
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            undefined,
          ]
        }
      >
        <View
          pointerEvents="none"
          style={
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.87)",
              "bottom": 0,
              "left": 0,
              "opacity": undefined,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": 1,
            }
          }
        />
        <View
          pointerEvents="none"
          style={
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.87)",
              "borderRadius": 100,
              "height": 200,
              "left": -100,
              "opacity": undefined,
              "position": "absolute",
              "top": -100,
              "transform": Array [
                Object {
                  "scale": undefined,
                },
              ],
              "width": 200,
              "zIndex": 1,
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onLayout={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      pointerEvents="box-only"
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
          },
          undefined,
          Object {
            "flex": 0,
            "minWidth": 200,
          },
        ]
      }
      testID={null}
    >
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "rgb(0, 0, 0)",
              "borderRadius": 21.6,
              "height": 43.2,
              "left": 2.3999999999999986,
              "opacity": undefined,
              "position": "absolute",
              "top": 2.3999999999999986,
              "transform": Array [
                Object {
                  "scale": undefined,
                },
              ],
              "width": 43.2,
              "zIndex": 1,
            }
          }
        />
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "height": 48,
                "justifyContent": "center",
                "width": 48,
              },
              Object {
                "height": 48,
                "width": 48,
              },
              undefined,
            ]
          }
        >
          <Text />
        </View>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "flex": 1,
              "marginLeft": 20,
            },
            undefined,
            undefined,
            false,
          ]
        }
      >
        auffällig, weil...
      </Text>
      <View
        pointerEvents="none"
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "overflow": "hidden",
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            undefined,
          ]
        }
      >
        <View
          pointerEvents="none"
          style={
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.87)",
              "bottom": 0,
              "left": 0,
              "opacity": undefined,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": 1,
            }
          }
        />
        <View
          pointerEvents="none"
          style={
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.87)",
              "borderRadius": 100,
              "height": 200,
              "left": -100,
              "opacity": undefined,
              "position": "absolute",
              "top": -100,
              "transform": Array [
                Object {
                  "scale": undefined,
                },
              ],
              "width": 200,
              "zIndex": 1,
            }
          }
        />
      </View>
    </View>
  </View>
  <View />
</View>
`;
